### PR TITLE
[CALCITE-4360] Apply SubstitutionRule first in top-down driven rule apply

### PR DESCRIPTION
**Pull request by chunweilei**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleQueue.java [hunk 1]
```diff
@@ -53,7 +53,7 @@ private void addMatch(VolcanoRuleMatch match, Deque<VolcanoRuleMatch> queue) {
       return;
     }
 
-    if (!planner.isSubstituteRule(match)) {
+    if (planner.isSubstituteRule(match)) {
       queue.addFirst(match);
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleQueue.java:57]: Here is the main change.

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleQueue.java:57]: Why it is still works before this patch ? Does that mean the place in the rule queue has no effect for SubstituteRule ?

Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleQueue.java:57]: The order of the rule does not have an impact on the final plan. But it has an impact on the efficiency of the optimizer.

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleQueue.java:57]: How much gains for this patch ?

File: core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java [hunk 2]
```diff
@@ -405,7 +406,7 @@ default boolean onProduce(RelNode node) {
   }
 
   /**
-   * Ensure that ExploreInputs are working on the correct input group.
+   * Ensures that ExploreInputs are working on the correct input group.
    * Currently, a RelNode's input may change since calcite may merge RelSets.
```
Comment (liyafan82) [inline core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java:410]: calcite -> Calcite?

File: core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java [hunk 3]
```diff
@@ -590,7 +591,7 @@ private RelNode convert(RelNode rel, RelSubset group) {
     return null;
   }
 
-  // check whether a node's lower bound is less than a RelSubset's upper bound
+  // Check whether a node's lower bound is less than a RelSubset's upper bound
```
Comment (liyafan82) [inline core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java:594]: Checks?

File: core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java [hunk 4]
```diff
@@ -1442,7 +1442,8 @@ public boolean isLogical(RelNode rel) {
   }
 
   /**
-   * Check whether a rule match is a substitute rule match.
+   * Checks whether a rule match is a substitute rule match.
```
Comment (liyafan82) [inline core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java:1445]: substitute rule -> {@link SubstituteRule}

**Comment by chunweilei** [comment]
I am sorry that I misunderstood the codes. Currently, the substitution rule is already applied first though it is not so straightforward. The process looks like:

1.put the non-substitution rule at the front and substitution rule at the end of the queue

2.get each rule from the queue in order from first to last and generate an ApplyRule task

3.push the ApplyRule task into the stack

As you can see, the ApplyRule(substitution) task is poped earlier than the ApplyRule(non-substitution) task from the stack. Thus substitution rule is executed first.

**Pull request closed without merging**